File: doc-examples/jdbc-example-python/src/main/python/example/PassengerRepository.py
from java.util.concurrent import CompletableFuture
from micronaut.data.annotation import Upsert
from micronaut.data.jdbc.annotation import JdbcRepository
from micronaut.data.model.query.builder.sql import Dialect
from micronaut.data.repository import CrudRepository

from example.Passenger import Passenger


# tag::upsert-repository[]
@JdbcRepository(dialect=Dialect.H2)
class PassengerRepository(CrudRepository[Passenger, int]):

    @Upsert(conflictsOn="email")
    def upsertByEmail(self, passenger: Passenger) -> Passenger: ...

    @Upsert(conflictsOn="email")
    def upsertAllByEmail(self, passengers: list[Passenger]) -> list[Passenger]: ...

    @Upsert(conflictsOn="email")
    def upsertByEmailFuture(self, passenger: Passenger) -> CompletableFuture[Passenger]: ...

    @Upsert(conflictsOn="email")
    def upsertAllByEmailFuture(self, passengers: list[Passenger]) -> CompletableFuture[list[Passenger]]: ...
# end::upsert-repository[]

File: doc-examples/jdbc-example-java/src/main/java/example/PassengerRepository.java
package example;

import io.micronaut.context.annotation.Requires;
import io.micronaut.data.annotation.Upsert;
import io.micronaut.data.jdbc.annotation.JdbcRepository;
import io.micronaut.data.model.query.builder.sql.Dialect;
import io.micronaut.data.repository.CrudRepository;

import java.util.List;
import java.util.concurrent.CompletableFuture;

@Requires(notEnv="oracle")
// tag::upsert-repository[]
@JdbcRepository(dialect = Dialect.H2)
public interface PassengerRepository extends CrudRepository<Passenger, Long> {

    @Upsert(conflictsOn = "email")
    Passenger upsertByEmail(Passenger passenger);

    @Upsert(conflictsOn = "email")
    List<Passenger> upsertByEmail(Iterable<Passenger> passengers);

    @Upsert(conflictsOn = "email")
    CompletableFuture<Passenger> upsertByEmailFuture(Passenger passenger);

    @Upsert(conflictsOn = "email")
    CompletableFuture<List<Passenger>> upsertByEmailFuture(Iterable<Passenger> passengers);
}
// end::upsert-repository[]
